src/Cake.NuGet/Install/Extensions/NuGetPackageManagerExtensions.cs

```diff
@@ -0,0 +1,258 @@
+            // get all dependencies of new target ids so that we can have all the required install actions.
+            var allowed = GetDependencies(newTargetIds, packagesToInstall, prunedAvailablePackages);
+
+            var nuGetProjectActions = new List<NuGetProjectAction>();
+            foreach (var packageToInstall in packagesToInstall)
+            {
+                // we should limit actions to just packages that are in the dependency set of the target we are installing
+                if (allowed.Contains(packageToInstall.Id))
+                {
+                    // find the package match based on identity
+                    var sourceDepInfo = prunedAvailablePackages.SingleOrDefault(p =>
+                        PackageIdentity.Comparer.Equals(p, packageToInstall));
+
+                    if (sourceDepInfo == null)
+                    {
+                        // this really should never happen
+                        throw new InvalidOperationException($"Package '{packageIdentity}' is not found");
+                    }
+
+                    nuGetProjectActions.Add(NuGetProjectAction.CreateInstallProjectAction(sourceDepInfo,
+                        sourceDepInfo.Source, nuGetProject));
+                }
+            }
+
+            return nuGetProjectActions;
+        }
+
+        private static async Task<SourceRepository> GetSourceRepositoryAsync(
+            PackageIdentity packageIdentity,
+            IEnumerable<SourceRepository> sourceRepositories,
+            ILogger logger,
+            CancellationToken token)
+        {
+            foreach (var sourceRepository in sourceRepositories)
+            {
+                var metadataResource = await sourceRepository.GetResourceAsync<MetadataResource>(token);
+                if (metadataResource != null)
+                {
+                    if (await metadataResource.Exists(packageIdentity, logger, token))
+                    {
+                        return sourceRepository;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private static HashSet<string> GetDependencies(
+            IEnumerable<string> targetIds,
+            IReadOnlyCollection<PackageIdentity> newListOfInstalledPackages,
+            IReadOnlyCollection<SourcePackageDependencyInfo> available)
+        {
+            var result = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var targetId in targetIds)
+            {
+                CollectDependencies(result, targetId, newListOfInstalledPackages, available, 0);
+            }
+            return result;
+        }
+
+        private static void CollectDependencies(HashSet<string> result,
+            string id,
+            IReadOnlyCollection<PackageIdentity> packages,
+            IReadOnlyCollection<SourcePackageDependencyInfo> available,
+            int depth)
+        {
+            // check if we've already added dependencies for current id
```

Comment:
Does `id` include version? How are multiple versions of the same dependency across different target IDs resolved - how do we know which one to get?

Reply:
Actually, this step is not required. Will remove it. The `packagesToInstall` in step-2 is the actual packages. This is only needed in sccenario's where we'd already have some packages installed and therefore some will need update and some should be installed.

Reply:
Gotcha